Structural data & JSON support

eras opened the pull request:
Hi,

This PR provides the ability to construct structural data and a way to dump it in JSON. It implements the particular support only for wt450.c for now, but others should be easy to merge if the first of these two commits gets in. Incorporating the patch for wt450 of course breaks scripts that depend on the output, but this would hopefully happen only once.

While this addresses the problems of issue #71 (modules just use fprintf), it doesn't provide CSV output because the schema of the data is not known up-front. It seems most if not all of the modules use a fixed format, so it might be a possible continuation idea for them to declare the fields they are going to use before the first data arrives. This would allow the data_acquired_handler-function to construct CSV from the same data.

This patch set may not be the most memory-efficient way to construct structured data, but it isn't perhaps that a horrible way either, given the low rate of data being transmitted. The interface is convenient:

```
data_t *data = data_make("time"       , DATA_TYPE_STRING, time_str,
                         "house code" , DATA_TYPE_INT, house_code,
                         "channel"    , DATA_TYPE_INT, channel,
                         "battery"    , DATA_TYPE_STRING, battery_low ? "LOW" : "OK",
                         "temp"       , DATA_TYPE_DOUBLE, temp,
                         "humidity"   , DATA_TYPE_INT, humidity,
                         NULL));
data_print_json(data, stdout);
data_free(data);
```

It supports not only strings and ints and stuff but also arrays and itself, so it should be suitable for generating arbitrary JSON. Perhaps in the future someone wants to write an XML or binary output module and this is possible as well ;).

It also comes with a test to exert the basic functions. Certainly not with great coverage. The code tries to deal with out-of-memory situations as well as rare they might be, but those paths are completely untested.

Keep up the good work!

merbanan commented:
I sort of had preferred to use YAML for the decoders. But how does a decoded packet look like ?

eras commented:
A decoded packet might look like (as the second patch implements it for WT450):

    {"time" : "2015-11-20 08:12:32", "house_code" : 1, "channel" : 1, "battery" : "OK", "temp" : 22.200001, "humidity" : 23}

Though the exact structure is easily changed. One idea might be to put "meta data" into its own structure to be filled by rtl_433 core code instead of the device drivers themselves. Then the output might look something like:

    {"time" : "2015-11-20 08:12:32", sensor:"WR450", data: { "house_code" : 1, "channel" : 1, "battery" : "OK", "temp" : 22.200001, "humidity" : 23}}

Extending this to support also YAML (and for example XML) would be quite easy, as the construction of the data and converting it to final representation are separate. Basically it would be implementing function data_print_yaml similar to the function data_print_json. There would be some refactoring involved to reduce code duplication in that case.

In practice I do believe JSON is:
- sufficient: JSON is more than able to express the data received from the sensors
- more practical: each JSON blob can be its own line, so there is no need to have additional message separators if the data is going to be just written to stdout. Newlines will do. People can use rtl_433 | while read msg; .. to do the processing as they have probably done so far :).
- de facto standard: it's going to be easier to use because JSON is much more widely supported.

merbanan commented:
The object should be linear, no nested structures.
I don't mind if the data from the decoders are json.
But what I really want is human readable output by default. And the option to transform it to cvs/plaintext.

merbanan commented:
Hmm, why is the temp value 22.200001 ?

eras commented:
Because of formulas involving floating point numbers :).

I can add support for formatted doubles.. They would be nice for the human readable part.

I'll also add support for an output that resembles the previous key-value output, but as currently every module just has their own fprintfs, their output quite likely won't conform to any single format..

merbanan commented:
We need to formalize and document the object type. Add a type member (sensor, remote, etc). Change house_code to id. Time to time_stamp. If possible make the human readable output to YAML. With that added I will accept and merge the patch.

merbanan commented:
Can you also supply some test signals of this device to the rtl_433_tests repo ?

merbanan commented:
And we need to rework the time stamp feature. The time stamp should be generated when the first signal is taken not in the decoders.

enavarro222 commented:
Hi,

I'm really interested in such a feature ! It will be much more easiest to pipe rtl_433 to an adhoc parsing script that push readings to a database.

For such usage, it is simpler to have only one line by sensor reading. In JSON it is possible. I don't think will be possible with YAML.

However a multi-lines YAML output may be nicer for a direct command line usage...

One can add an option to choose output format ?

I hope to see that soon in rtl_433 (by the way: thanks for this soft !)
I may help to change some decoder to this new output system.

merbanan commented:
@enavarro222 you could help by adding sensor recordings to the rtl_433_tests repository.

rct commented:
First, I think it would be helpful to move the discussion of reworking rtl_433's architecture to the rtl_433 google group.  https://groups.google.com/d/forum/rtl_433

I have a strong preference for single line output formats.  This lets me easily manipulate the output especially for diagnosis using tools such as grep to pull lines I need out of the log file.  I have currently have over 25 devices that are transmitting on 433 mhz - Lacrosse TX, DSC security contacts, and 3 types of Acurite devices.

What advantages does YAML have over JSON?  I could see that it would be more readable for a low volume of output.  However, if the RF environment is busy, the output will scroll off the screen too quickly and become a hindrance.

Finally, given the number of existing devices (now 40 with my last pull request) and the scripts/systems we all have for putting the output to use, I think it an output switch is required before converting existing devices.

rct commented:
FYI, the Desert Home fork of rtl_433, emits JSON for the Acurite 5n1 weather station:

{"sensorId":{"SID":"92","t":"1434292619"},"channel":{"CH":"A","t":"1434292619"},"messageCaught":{"MC":"0","t":"1434292619"},"battLevel":{"BAT":"7","t":"1434292619"},"windSpeed":{"WS":"3.5","t":"1434292619"},"windDirection":{"WD":"NNE","t":"1434292619"},"temperature":{"T":"89.8","t":"1434292601"},"humidity":{"H":"27","t":"1434292601"},"rainCounter":{"RC":"445","t":"1434292619"}}

The 5n1 weather station sends two different message types. so the above output has aggregated them into a single message, which is why each element is time stamped.  

Personally, I think the output should be more raw. I'd rather see the two individual messages that the device sends separately, rather than have rtl_433 keep state and do some processing that might obscure information that downstream processing could use.

eras commented:
I added human-readable key-value output, which looks like this (implemented for wt450 only so far):

	% rtl_433 -R 33 2>/dev/null
	2015-11-21 18:01:51 WT450 sensor: House Code 1, Channel 1, Battery OK, Temperature 23.20 C, Humidity 22 %
	2015-11-21 18:01:57 WT450 sensor: House Code 1, Channel 2, Battery OK, Temperature 5.25 C, Humidity 35 %
	2015-11-21 18:01:58 WT450 sensor: House Code 1, Channel 2, Battery OK, Temperature 5.28 C, Humidity 35 %

And the JSON output looks like this:

	% rtl_433 -R 33 -j 2>/dev/null
	{"time" : "2015-11-21 18:02:51", "label" : "WT450 sensor", "house_code" : 1, "channel" : 1, "battery" : "OK", "temp" : 23.400000, "humidity" : 22}
	{"time" : "2015-11-21 18:02:51", "label" : "WT450 sensor", "house_code" : 1, "channel" : 1, "battery" : "OK", "temp" : 23.420000, "humidity" : 22}
	{"time" : "2015-11-21 18:02:51", "label" : "WT450 sensor", "house_code" : 1, "channel" : 1, "battery" : "OK", "temp" : 23.450001, "humidity" : 22}

And finally the CSV output:

	% rtl_433 -R 33 -c 2>/dev/null
	time;label;house_code;channel;battery;temp;humidity
	2015-11-21 18:03:25;WT450 sensor;1;1;OK;22.639999;21
	2015-11-21 18:03:25;WT450 sensor;1;1;OK;22.660000;21
	2015-11-21 18:03:26;WT450 sensor;1;1;OK;22.690001;21

Why no truncated decimals? Because there is only one formatting function for 'human readable' output, and as it contains also the "C" suffix, I decided not to use the same formatter for the other cases. This is all a driver needs to do:

	data = data_make(
		"",           "time" ,       DATA_STRING, time_str,
		"",	      "label" ,      DATA_STRING, "WT450 sensor",
		"House Code", "house_code",  DATA_INT, house_code,
		"Channel",    "channel",     DATA_INT, channel,
		"Battery",    "battery",     DATA_STRING, battery_low ? "LOW" : "OK",
		"Temperature","temp",        DATA_FORMAT, "%.02f C", DATA_DOUBLE, temp,
		"Humidity",   "humidity",    DATA_FORMAT, "%u %%", DATA_INT, humidity,
		NULL);
	data_acquired_handler(data);

And also for the benefit of CSV support:

	static char *output_fields[] = { "time", "label", "house_code", "channel", "battery", "temp", "humidity", NULL };
	
	r_device wt450 = {
	..other fields..
	   .json_callback = &wt450_callback,
	   .fields        = output_fields
	};

Pretty bearable.

If there's something to add (ie. common fields, they can be done in the rtl_433.c) in the data_acquired_handler function. data doesn't yet support adding additional fields to it (so hierarchies would be eaiser ;)), but if some format is decided upon, that's probably easy to implement to apply for all protocols.

rct commented:
@eras - this all looks like a great start!

As I said previously,  I think it would be helpful to move the discussion to the rtl_433 google group. https://groups.google.com/d/forum/rtl_433.

A few comments:
- Labels such as "temp" are ambiguous, is it C of F?   Off the top of my head, I currently don't think we need a separate unit type for each field. I think the label should be sufficient, tempc vs. tempf, wind-mph, wind-mps, etc.
- The device's "native" units should be reported at a minimum. If a device reports in Fahrenheit, it shouldn't be converted to Celsius and back to Fahrenheit again to avoid loss of precision especially after rounding has occurred to make pretty printed floats. If it is decided that message formats should always include metric units, then the original units should be included if they weren't metric.
- The device manufacturer and model number should both be included, preferably as separate fields. Downstream parsers might want want to identify such as Acurite and Oregon Scientific.

Temperature and humidity sensors are pretty straightforward data wise. It would be helpful to look at a wider range of devices before baking the protocol too much. Even with those there is a bit of a range of identifiiers, serial number, channel, house code, etc.

jcarduino commented:
Is it an option to make a switch for metric so that switch indicated inch f c meters etc

Mvg Johan Corstjens

Verstuurd via mobile 

> Op 21 nov. 2015 om 19:43 heeft rct <notifications@github.com> het volgende geschreven:
> 
> @eras - this all looks like a great start!
> 
> As I said previously, I think it would be helpful to move the discussion to the rtl_433 google group. https://groups.google.com/d/forum/rtl_433.
> 
> A few comments:
> 
> Labels such as "temp" are ambiguous, is it C of F? Off the top of my head, I currently don't think we need a separate unit type for each field. I think the label should be sufficient, tempc vs. tempf, wind-mph, wind-mps, etc.
> The device's "native" units should be reported at a minimum. If a device reports in Fahrenheit, it shouldn't be converted to Celsius and back to Fahrenheit again to avoid loss of precision especially after rounding has occurred to make pretty printed floats. If it is decided that message formats should always include metric units, then the original units should be included if they weren't metric.
> The device manufacturer and model number should both be included, preferably as separate fields. Downstream parsers might want want to identify such as Acurite and Oregon Scientific.
> Temperature and humidity sensors are pretty straightforward data wise. It would be helpful to look at a wider range of devices before baking the protocol too much. Even with those there is a bit of a range of identifiiers, serial number, channel, house code, etc.
> 
> —
> Reply to this email directly or view it on GitHub.
>

merbanan commented:
@rct I think that decoders should pass down native data from the sensors. The receiving layer can normalize the data later if selected. So if there is temperature reading in Celsius a parameter with unit Celsius should be used. The normalizing system could then transform it to Kelvin later.

eras commented:
Btw, adding units could work like this:

	data = data_make(
	    "Temperature", "temperature", DATA_UNIT, "C", DATA_DOUBLE, 12.0.
	    NULL);

Some aspects of this approach:

- The JSON output is not quite self-evident. Some options:

        { "temperature_C" : 12.0 } <- my favorite I think
	{ "temperature" : { "value": 12.0, "unit" : "C" } }
	{ "values": [{name : "temperature", "value" : 12.0, "unit" : "C" }] }

- works with the current key-value-output format easily (just append the unit after the value)
- it would be immediately useful even without an elaborate subsystem to do conversions between units
- it's opt in: existing code could be converted as-is without unit-support
- the key names are ambigious in the sense that underscore can be a unit separator as well as a symbol in the original label

merbanan commented:
@eras CSV should be separated by comma.

Variable names with the included unit is preferred. It is not the job of the decoders to normalized data. They should just report what they decode. The JSON structure should be possible to modify by a normalizing step. So that Fahrenheit can be converted to Celsius if needed. This can wait until later, but we need the decoders to use different names for units so they can be normalized later. temperature_C would be fine in this case.

The human readable output should be placed on several lines for easy parsing by the eyes.

Regarding id and channel, those parameters need to be separate as channel is a fixed variable parameter. While id can be a random value that i generated when batteries are inserted. Maybe we need to expose it as random_id and fixed_id. The main idea behind all this is that an event parser system should be able to do a 1:1 mapping from generated events.

eras commented:
Rewrote the whole bunch for cleaner commit history ;).

- reordered the arguments so pretty_key can be omitted (so it defaults to the key then)
- renamed house_code->id, label->model, temp->temperature_C
- used , for CSV separator (but really, there's no standards and ; can be mighty useful if strings (names) contain comma themselves)
- changed the command line switches -j for json and -c for csv to -F kv|json|csv defaulting to kv
- added support for waveman as well. This changes how the default waveman output looks like, though, as it is completely different from what wt450 did..

merbanan commented:
Ok, I think I am happy enough with the patch. I will leave it for some time before I merge it.

@rct any comments ? I think that we have almost we need. The details can be ironed out later. I think this is good step forward.

eras commented:
Oops, I had somehow broken the waveman output slightly while adjusting the kv output for wr450. For the record, it now looks like this (line-oriented or YAML output is still unimplemented):

    model Waveman Switch Transmitter, id F, channel 1, button 4, state off

merbanan commented:
Ok, this looks good. Thanks for the patch.
